dao/src/main/java/org/wso2/testgrid/dao/repository/TestPlanRepository.java

```diff
@@ -133,4 +134,91 @@ public TestPlan findByPrimaryKey(String id) throws TestGridDAOException {
                     " query [", queryStr, "]"), e);
         }
     }
+
+    /**
+     * This method returns the last failed {@link TestPlan} for a given product.
+     *
+     * @param product A Product object being queried.
+     * @return instance of a {@link TestPlan} representing the last failed test plan.
+     * @throws TestGridDAOException when an error occurs while executing the native sql query.
+     */
+    public TestPlan getLastFailure(Product product) throws TestGridDAOException {
+        String sql = "SELECT  t.* from test_plan t INNER JOIN (SELECT tp.infra_parameters," +
+                "max(tp.modified_timestamp) AS time, dp.name FROM test_plan tp INNER JOIN " +
+                "deployment_pattern dp ON tp.DEPLOYMENTPATTERN_id=dp.id AND tp.status='FAIL' " +
+                "AND  tp.DEPLOYMENTPATTERN_id IN (SELECT id FROM deployment_pattern WHERE " +
+                "PRODUCT_id='" + product.getId() + "') GROUP BY tp.infra_parameters,dp.name) AS x ON " +
+                "t.infra_parameters=x.infra_parameters AND t.modified_timestamp=x.time ORDER BY time DESC LIMIT 1";
+        try {
+            return (TestPlan) entityManager.createNativeQuery(sql, TestPlan.class).getSingleResult();
+        } catch (Exception e) {
```

Comment:
We usually put the try block for the entire logic, not just for the line where the exception is thrown.

And, shall we catch a specific exception here instead of Exception?